generated/nidaqmx/task.py

```diff
@@ -759,6 +677,102 @@ def read(self, number_of_samples_per_channel=NUM_SAMPLES_UNSET,
+                CtrFreq(freq=f, duty_cycle=d) for f, d in zip(frequencies, duty_cycles)
+            ]
+
+        elif meas_type == UsageTypeCI.PULSE_TIME:
+            high_times = numpy.zeros(array_shape, dtype=numpy.float64)
+            low_times = numpy.zeros(array_shape, dtype=numpy.float64)
+
+            _, _, samples_read = self._interpreter.read_ctr_time(
+                self._handle, number_of_samples_per_channel, timeout,
+                FillMode.GROUP_BY_CHANNEL.value, high_times, low_times)
+            data = [CtrTime(high_time=h, low_time=l) for h, l in zip(high_times, low_times)]
+
+        elif meas_type == UsageTypeCI.PULSE_TICKS:
+            high_ticks = numpy.zeros(array_shape, dtype=numpy.uint32)
+            low_ticks = numpy.zeros(array_shape, dtype=numpy.uint32)
+
+            _, _ , samples_read = self._interpreter.read_ctr_ticks(
+                self._handle, number_of_samples_per_channel, timeout,
+                FillMode.GROUP_BY_CHANNEL.value, high_ticks, low_ticks)
+            data = [CtrTick(high_tick=h, low_tick=l) for h, l in zip(high_ticks, low_ticks)]
+
+        else:
+            assert False, f"{meas_type} is not a counter pulse measurement type."
+        
+        if num_samples_not_set and array_shape == (1,):
+            return data[0]
+
+        # Counter pulse measurements should not have N channel versions.
+        #
+        # https://github.com/ni/nidaqmx-python/issues/498 - Missing support for
+        # multi-channel counter reads in stream_readers and task
+        if samples_read != number_of_samples_per_channel:
+            assert number_of_channels == 1
+            return data[:samples_read]
+
+        return data
+
+    def _read_power(
+        self,
+        array_shape: Tuple[int, ...],
+        number_of_channels: int,
+        number_of_samples_per_channel: int,
+        timeout: float,
+    ) -> Union[PowerMeasurement, List[PowerMeasurement], List[List[PowerMeasurement]]]:
+        voltages = numpy.zeros(array_shape, dtype=numpy.float64)
+        currents = numpy.zeros(array_shape, dtype=numpy.float64)
+
+        _, _, samples_read = self._interpreter.read_power_f64(
+            self._handle, number_of_samples_per_channel, timeout,
+            FillMode.GROUP_BY_CHANNEL.value, voltages, currents)
+
+        if number_of_channels > 1:
+            if number_of_samples_per_channel == 1:
+                # n channel, 1 sample
+                return [
+                    PowerMeasurement(voltage=v, current=i)
+                    for v, i in zip(voltages, currents)
+                ]
+            else:
+                # n channel, n samples
+                return [
+                    [
+                        PowerMeasurement(voltage=v, current=i)
+                        for v, i in zip(voltages[channel_index], currents[channel_index])
+                    ]
+                    for channel_index in range(number_of_channels)
+                ]
```

Comment:
Shouldn't this limit to `[:samples_read]`?

Reply:
Good news: I added a test for this. https://github.com/ni/nidaqmx-python/pull/530

Bad news: I found two other issues besides this one.
- #527 
- #528 
- #529 
@zhindes @hongloonwong